### ISSUE TITLE
LFRic reductions use default precision

### ISSUE THREAD

**Issue by rupertford** [stfc/PSyclone#1570]
In PSyclone there is an assumption that the precision of a scalar reduction variable will always be the default value (currently r_def).

This needs discussing with the Met Office to determine their requirements and see whether this assumption needs changing.

**Comment by rupertford** [comment]
@christophermaynard has just found that his branch does not work with PSyclone due to him declaring a reduction scalar with precision `r_solver`.

**Comment by rupertford** [comment]
Here is an example of generated code to help make it clear what the issues are ...

```
REAL(KIND=r_def), intent(out) :: asum
...
TYPE(scalar_type) global_sum
...
asum = 0.0_r_def
!
DO df=1,f1_proxy%vspace%get_last_dof_owned()
    asum = asum + f1_proxy%data(df)*f2_proxy%data(df)
END DO
global_sum%value = asum
asum = global_sum%get_sum()
```

**Comment by rupertford** [comment]
If the scalar reduction variable `asum` was declared as `r_solver` then the global sum (implemented in `global_sum`) would still be `r_def` as `scalar_type` uses `r_def` so any global sums would be done with `r_def` precision.

However, perhaps there would be some performance gain in performing local sums in lower precision.

I think the Met Office need to decide what they want here. @christophermaynard, care to comment?

**Comment by TeranIvy** [comment]
Note that we also have `r_tran_field_type` to support the mixed precision transport, see issue #1632.

**Comment by TeranIvy** [comment]
Well, we need this so I'll talk with Mike about the infrastructure support.

**Comment by TeranIvy** [comment]
The associated LFRic infrastructure support ticket is https://code.metoffice.gov.uk/trac/lfric/ticket/3401.

**Comment by TeranIvy** [comment]
From #2342: _When local reductions are performed on fields, there is currently no guarantee that the precision of the field is converted to that of the local reduction variable (`l_asum` e.g.). To fix this, a precision comparison should be done between the field and the local reduction variable. If the precisions do not match, the field's data should be converted to the precision of the local sum variable._

**Comment by TeranIvy** [comment]
**Note** the related issue to introduce GlobalReduction and the relevant API sub-classes, #2381.

**Comment by arporter** [comment]
@mo-alistairp can you remember why your PR #3382 was only a 'towards' this issue rather than 'closes'?

**Comment by mo-alistairp** [comment]
> [@mo-alistairp](https://github.com/mo-alistairp) can you remember why your PR [#3382](https://github.com/stfc/PSyclone/pull/3382) was only a 'towards' this issue rather than 'closes'?

No, it seems that it should have been a 'closes' with #3382. I'll close this issue
